File: Vec/dotprod.m
function prod = dotprod(u, v, prod, nthreads)
%DOTPROD  computes dot prod of column vectors.
%
%      prod = dotprod(u, v)
% computes the dot product in serial. Both u and v are assumed to be
% column vectors. If they have multiple columns, then compute the dot
% product of each pair of columns and save the answer into columns of prod.
%
%      prod = dotprod(u, v, prod, nthreads)
% computes the dot product using multicores, as controlled by nthreads.
% Note that size(prod,1) should be >= nthreads to facilitate
% concurrency. Otherwise, it will use up to size(prod,1) threads.
%
% If nthreads is empty, then it computes prod=u'*v on a local chunk per
% thread. This assumes that a parallel region has already been started.
% In this mode, prod must be a variable shared by all threads in the team,
% and size(prod,1) should be >= nthreads to facilitate concurrency.
% Otherwise, it will use up to size(prod,1) threads.
%
% In multithread, this is a collective function.
% In multithread mode, there is an implicit barrier at the beginning and
% at the end of this function.
%
% Note: The compiled version requires u and v to be double precision.
%       The M file works for both single and double precision.
%
% See also momp_Control MPI_Comm

%#codegen -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),
%#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0])}
%#codegen dotprod_ser -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}

coder.inline('never');

if size(u,1)~=size(v,1) || size(u,2)~=size(v,2)
    momp_begin_master
    m2c_error('dotprod:IncorrectSize', 'Dimensions for u and v must match.');
    momp_end_master
end

if nargin<3
    prod=nullcopy(zeros(1,size(u,2)));
elseif size(prod,1)<1 || size(prod,2)~=size(u,2)
    momp_begin_master
    m2c_error('dotprod:IncorrectSize', 'prod and u must have the same number of columns.');
    momp_end_master
end

ismt = nargin>3 && momp_get_num_threads>1;

if nargin>3 && ~isempty(nthreads)
    %% Declare parallel region
    if ~momp_get_nested && momp_get_num_threads>1 && nthreads(1)>1
        momp_begin_master
        m2c_warn('dotprod:NestedParallel', ...
            'You are trying to use nested parallel regions, but nested parallelism is not enabled.');
        momp_end_master
    end
    
    momp_begin_parallel; momp_clause_default('shared');
    momp_clause_num_threads(nthreads(1));
    
    %% Compute prod_local=u'*v
    prod = dotprod_partial(u, v, prod, '');
    momp_barrier
    
    prod = accu_partsum(prod);    
    %% End parallel region
    momp_end_parallel;
elseif ismt
    %% Compute prod_local=u'*v
    prod = dotprod_partial(u, v, prod, '');
    momp_barrier
    
    prod = accu_partsum(prod);
    momp_barrier
else
    prod = dotprod_partial(u, v, prod);
end

function prod = dotprod_partial(u, v, prod, varargin)
% Compute partial dot product with each therad

if nargin>3
    coder.inline('never');
    ind = momp_get_thread_num+1;
    [istart, iend] = get_local_chunk(int32(size(u,1)));
else
    ind = int32(1);
    istart = int32(1); iend = int32(size(u, 1));
end

for k=1:int32(size(u,2))
    prod(ind, k) = 0.;
    for i=istart:iend
        prod(ind, k) = prod(ind, k) + u(i, k) * v(i, k);
    end
end

function prod = accu_partsum(prod)
% Accumulate the partial sums of all columns
coder.inline('always');

[istart, iend] = get_local_chunk(int32(size(prod,2)));
n = min(momp_get_num_threads, int32(size(prod,1)));

for k=istart:iend
    for j=2:n
        prod(1,k) = prod(1,k) + prod(j,k);
    end
end

function test %#ok<DEFNU>
%!test
%! if ~exist(['dotprod.' mexext], 'file')
%!    m=200;
%! else
%!    m=1000000;
%! end
%! tic; u =rand(m,2); v=rand(m,2);
%! fprintf(1, '\n\tGenerated random matrix in %g seconds\n', toc);
%! tic; b0 = [u(:,1)'*v(:,1), u(:,2)'*v(:,2)];
%! fprintf(1, '\tComputed reference solution in %g seconds\n', toc);
%! fprintf(1, '\tTesting serial: ');
%! tic; b1 = dotprod(u, v);
%! fprintf(1, 'Done in %g seconds\n ', toc);
%! assert(norm(b0-b1)/norm(b0)<=1.e-6);

%! for nthreads=int32([1 2 4 8])
%!     if nthreads>momp_get_max_threads; break; end
%!     fprintf(1, '\tTesting %d thread(s): ', nthreads);
%!     b2 = zeros(nthreads,2);
%!     tic; b2 = dotprod(u, v, b2, nthreads);
%!     fprintf(1, 'Done in %g seconds\n ', toc);
%!     assert(norm(b0-b2(1,:))/norm(b0)<=1.e-6);
%! end

File: Vec/codegen/lib/dotprod/dotprod.c
#include "dotprod.h"
#include "omp.h"
#include "m2c.h"

static void accu_partsum(emxArray_real_T *prod);
static void b_dotprod(const emxArray_real_T *u, const emxArray_real_T *v,
                      emxArray_real_T *prod);
static void b_m2c_error(void);
static void dotprod_partial(const emxArray_real_T *u, const emxArray_real_T *v,
  emxArray_real_T *prod);
static void m2c_error(void);
static void m2c_warn(void);
static double rt_roundd(double u);
static void accu_partsum(emxArray_real_T *prod)
{
  int nthreads;
  int thr_id;
  int istart;
  double y;
  int iend;
  double b_y;
  int varargin_1;
  int chunk;
  int n;
  int b_remainder;
  int k;
  int c_remainder;
  int j;
  nthreads = omp_get_num_threads();
  if (nthreads == 1) {
    istart = 0;
    iend = prod->size[1];
  } else {
    thr_id = omp_get_thread_num();
    y = (double)prod->size[1] / (double)nthreads;
    if (y < 0.0) {
      b_y = ceil(y);
    } else {
      b_y = floor(y);
    }

    chunk = (int)rt_roundd(b_y);
    b_remainder = prod->size[1] - nthreads * chunk;
    if (b_remainder <= thr_id) {
      c_remainder = b_remainder;
    } else {
      c_remainder = thr_id;
    }

    istart = thr_id * chunk + c_remainder;
    iend = (istart + chunk) + (thr_id < b_remainder);
  }

  varargin_1 = omp_get_num_threads();
  if (varargin_1 <= prod->size[0]) {
    n = varargin_1;
  } else {
    n = prod->size[0];
  }

  for (k = istart; k + 1 <= iend; k++) {
    for (j = 2; j <= n; j++) {
      prod->data[prod->size[0] * k] += prod->data[(j + prod->size[0] * k) - 1];
    }
  }
}

static void b_dotprod(const emxArray_real_T *u, const emxArray_real_T *v,
                      emxArray_real_T *prod)
{
  int i0;
  int k;
  int i;
  if ((u->size[0] != v->size[0]) || (u->size[1] != v->size[1])) {

#pragma omp master

    M2C_BEGIN_REGION(/*omp master*/)

    m2c_error();

    M2C_END_REGION(/*omp master*/)

  }

  i0 = prod->size[0] * prod->size[1];
  prod->size[0] = 1;
  prod->size[1] = u->size[1];
  emxEnsureCapacity((emxArray__common *)prod, i0, (int)sizeof(double));
  for (k = 0; k + 1 <= u->size[1]; k++) {
    prod->data[prod->size[0] * k] = 0.0;
    for (i = 0; i + 1 <= u->size[0]; i++) {
      prod->data[prod->size[0] * k] += u->data[i + u->size[0] * k] * v->data[i +
        v->size[0] * k];
    }
  }
}

static void b_m2c_error(void)
{
  M2C_error("dotprod:IncorrectSize",
            "prod and u must have the same number of columns.");
}

static void dotprod_partial(const emxArray_real_T *u, const emxArray_real_T *v,
  emxArray_real_T *prod)
{
  int n;
  int nthreads;
  int thr_id;
  int istart;
  double y;
  int iend;
  double b_y;
  int k;
  int chunk;
  int b_remainder;
  int i;
  int c_remainder;
  n = omp_get_thread_num();
  nthreads = omp_get_num_threads();
  if (nthreads == 1) {
    istart = 0;
    iend = u->size[0];
  } else {
    thr_id = omp_get_thread_num();
    y = (double)u->size[0] / (double)nthreads;
    if (y < 0.0) {
      b_y = ceil(y);
    } else {
      b_y = floor(y);
    }

    chunk = (int)rt_roundd(b_y);
    b_remainder = u->size[0] - nthreads * chunk;
    if (b_remainder <= thr_id) {
      c_remainder = b_remainder;
    } else {
      c_remainder = thr_id;
    }

    istart = thr_id * chunk + c_remainder;
    iend = (istart + chunk) + (thr_id < b_remainder);
  }

  for (k = 0; k + 1 <= u->size[1]; k++) {
    prod->data[n + prod->size[0] * k] = 0.0;
    for (i = istart; i + 1 <= iend; i++) {
      prod->data[n + prod->size[0] * k] += u->data[i + u->size[0] * k] * v->
        data[i + v->size[0] * k];
    }
  }
}

static void m2c_error(void)
{
  M2C_error("dotprod:IncorrectSize", "Dimensions for u and v must match.");
}

static void m2c_warn(void)
{
  M2C_warn("dotprod:NestedParallel",
           "You are trying to use nested parallel regions, but nested parallelism is not enabled.");
}

static double rt_roundd(double u)
{
  double y;
  if (fabs(u) < 4.503599627370496E+15) {
    if (u >= 0.5) {
      y = floor(u + 0.5);
    } else if (u > -0.5) {
      y = 0.0;
    } else {
      y = ceil(u - 0.5);
    }
  } else {
    y = u;
  }

  return y;
}

void dotprod(const emxArray_real_T *u, const emxArray_real_T *v, emxArray_real_T
             *prod, const emxArray_int32_T *nthreads)
{
  int n;
  int k;
  int i;
  if ((u->size[0] != v->size[0]) || (u->size[1] != v->size[1])) {

#pragma omp master

    M2C_BEGIN_REGION(/*omp master*/)

    m2c_error();

    M2C_END_REGION(/*omp master*/)

  }

  if ((prod->size[0] < 1) || (prod->size[1] != u->size[1])) {

#pragma omp master

    M2C_BEGIN_REGION(/*omp master*/)

    b_m2c_error();

    M2C_END_REGION(/*omp master*/)

  }

  n = omp_get_num_threads();
  if (!(nthreads->size[0] == 0)) {
    n = omp_get_nested();
    if (!(n != 0)) {
      n = omp_get_num_threads();
      if ((n > 1) && (nthreads->data[0] > 1)) {

#pragma omp master

        M2C_BEGIN_REGION(/*omp master*/)

        m2c_warn();

        M2C_END_REGION(/*omp master*/)

      }
    }

#pragma omp parallel default(shared) num_threads(nthreads->data[0])
    M2C_BEGIN_REGION(/*omp parallel*/)

    dotprod_partial(u, v, prod);

#pragma omp barrier

    accu_partsum(prod);

    M2C_END_REGION(/*omp parallel*/)

  } else if (n > 1) {
    dotprod_partial(u, v, prod);

#pragma omp barrier

    accu_partsum(prod);

#pragma omp barrier

  } else {
    for (k = 0; k + 1 <= u->size[1]; k++) {
      prod->data[prod->size[0] * k] = 0.0;
      for (i = 0; i + 1 <= u->size[0]; i++) {
        prod->data[prod->size[0] * k] += u->data[i + u->size[0] * k] * v->data[i
          + v->size[0] * k];
      }
    }
  }
}

void dotprod_initialize(void)
{
}

void dotprod_ser(const emxArray_real_T *u, const emxArray_real_T *v,
                 emxArray_real_T *prod)
{
  b_dotprod(u, v, prod);
}

void dotprod_terminate(void)
{
}

void emxInitArray_int32_T(emxArray_int32_T **pEmxArray, int numDimensions)
{
  emxInit_int32_T(pEmxArray, numDimensions);
}

void emxInitArray_real_T(emxArray_real_T **pEmxArray, int numDimensions)
{
  emxInit_real_T(pEmxArray, numDimensions);
}
